File: src/juggle_sort.java
/*
 * Juggle Sort - Java Version
 * @sayanwala117 / Jaspreet Jawanda
 * Date Modified: 2025-03-17
 * 
 * Backstory:
 * Created this sorting algorithm in class while the professor was teaching merge sort.
 * Although lowkey cooked for Data Structures and Algorithms, I thought I'd gain aura from making
 * a sorting algorithm. That did not happen—I still got no aura.
 * 
 * What does it do?
 * Finds the min and max, removes them, then adds them to a new sorted list, and removes them from 
 * the original list. This can be more efficient by using one list and avoiding the remove function.
 * Calling it Juggle Sort because it "juggles" the min and max over and over again.
 * 
 * This is very inefficient as it currently stands. If you use it and think, 
 * "What was going through this guy’s mind?" be assured the winter semester depression is kicking in.
 * 
 * Notes:
 * - The current implementation still uses `.remove()`, making it inefficient.
 * - The recursive structure remains the same as the original Python version.
 * - It should theoretically work for a linked list as well—if you manage to implement that, let me know.
 */

import java.util.ArrayList;
import java.util.List;

public class JuggleSort {

    // Finds the minimum value in a list
    public static int findMin(List<Integer> lst) {
        int minVal = lst.get(0);
        int size = lst.size();

        for (int i = 0; i <= size / 2; i++) {
            if (lst.get(i) < minVal) {
                minVal = lst.get(i);
            }
            if (lst.get(size - 1 - i) < minVal) {
                minVal = lst.get(size - 1 - i);
            }
        }
        return minVal;
    }

    // Finds the maximum value in a list
    public static int findMax(List<Integer> lst) {
        int maxVal = lst.get(0);
        int size = lst.size();

        for (int i = 0; i <= size / 2; i++) {
            if (lst.get(i) > maxVal) {
                maxVal = lst.get(i);
            }
            if (lst.get(size - 1 - i) > maxVal) {
                maxVal = lst.get(size - 1 - i);
            }
        }
        return maxVal;
    }

    // Sorts the list using Juggle Sort
    private static List<Integer> juggleSort(List<Integer> lst, List<Integer> sortedLst, int index) {
        if (sortedLst == null) {
            sortedLst = new ArrayList<>();
            for (int i = 0; i < lst.size(); i++) {
                sortedLst.add(null);
            }
        }

        if (lst.isEmpty()) {
            return sortedLst;
        }

        int minVal = findMin(lst);
        int maxVal = findMax(lst);

        sortedLst.set(index, minVal);
        sortedLst.set(sortedLst.size() - 1 - index, maxVal);

        lst.remove((Integer) minVal);
        if (minVal != maxVal) {
            lst.remove((Integer) maxVal);
        }

        return juggleSort(lst, sortedLst, index + 1);
    }

    public static List<Integer> juggleSort(List<Integer> lst) {
        return juggleSort(new ArrayList<>(lst), null, 0);
    }
}

File: src/juggle_sort.py
"""

Juggle sort
@sayanwala117 / Jaspreet Jawanda
Date Modified: 2025 - 03 - 17

Back story:
Created this sorting algorithm in class as the professor was teaching us merge sort
althogh lowkey cooked for Data structures and algoritms thought Id gain aura from making an 
sorting algortim, that did not happen I still got no aura.

what does it do?
Finds min and max, removes them then adds them to the new sorted list, and then it removes it from orginal list. 
This can be more efficent by using one list and not using the remove function. Calling it juggle sort because it juggles 
the min and max over and over again. This is very inefficent as it currently stands. If you use it and think "what was going through
this guys mind", be assured the winter semester depression is kicking in.  
    
"""

def find_min(lst):
    """
    _summary_
    find the minimum value in a list

    Args:
        lst (list): the orginal list

    Returns:
        min_val = int
    """
    min_val = lst[0]
    index = 0
    for i in range((len(lst) // 2) + 1):
        if lst[i] < min_val:
            min_val = lst[i]

        if lst[len(lst) - 1 - i] < min_val:
            min_val = lst[len(lst) - 1 - i]
            
    return min_val

def find_max(lst):
    """
    _summary_
    find the max value in a list

    Args:
        lst (list): the orginal list

    Returns:
        max_val = int
    """
    max_val = lst[0]
    
    for i in range((len(lst) // 2) + 1):
        if lst[i] > max_val:
            max_val = lst[i]
            
        if lst[len(lst) - 1 - i] > max_val:
            max_val = lst[len(lst) - 1 - i]

    return max_val

def sort(lst, sorted_lst=None, index=0):
    """_summary_
    recursivly calls and finds the min and max after remvoing the min and max in orginal. 
    also forgot to add it also creates a empty sorted list with none type. pretty sure this would
    work for a linked list too, if you manage to implement juggle sort for some like dat lemme know
    
    for my non programing buddies, this is meat of the sorting algortim.

    Args:
        lst (_type_): _description_
        sorted_lst (_type_, optional): _description_. Defaults to None.
        index (int, optional): _description_. Defaults to 0.

    Returns:
        _type_: _description_
    """
    if sorted_lst is None:
        sorted_lst = [None] * len(lst)
    
    if not lst:
        return sorted_lst
    
    min_val = find_min(lst)
    max_val = find_max(lst)
    
    sorted_lst[index] = min_val
    sorted_lst[-(index + 1)] = max_val
    
    lst.remove(min_val)
    if min_val != max_val:
        lst.remove(max_val)
    
    return sort(lst, sorted_lst, index + 1)

